crypto-sodium/lib/Crypto/Hash/Internal.hs

```diff
@@ -0,0 +1,60 @@
+-- SPDX-FileCopyrightText: 2021 Serokell
+--
+-- SPDX-License-Identifier: MPL-2.0
+
+{-# LANGUAGE ExplicitNamespaces, TypeOperators, TypeFamilies #-}
+{-# OPTIONS_GHC -Wno-redundant-constraints #-}
+
+-- | Internals of @crypto_generichash@.
+module Crypto.Hash.Internal
+  ( HashBlake2b
+  , blake2b
+  ) where
+
+import Prelude hiding (length)
+
+import Data.ByteArray (ByteArray, ByteArrayAccess, length, withByteArray)
+import Data.ByteArray.Sized (SizedByteArray, allocRet)
+import Data.Proxy (Proxy (Proxy))
+import Foreign.Ptr (nullPtr)
+import GHC.TypeNats (KnownNat, natVal, type (<=))
+
+import qualified Libsodium as Na
+
+-- | Hash returned by 'blake2b'.
+--
+-- This type is parametrised by hash size in bytes and the actual data type
+-- that contains bytes. This can be, for example, a @ByteString@.
+--
+-- Length must be between 16 and 64 bytes.
+type HashBlake2b len a = SizedByteArray len a
+
+-- | Hash a message using BLAKE2b.
+blake2b
+  ::  forall pt key hashBytes len.
+      ( ByteArrayAccess pt
+      , ByteArrayAccess key
+      , ByteArray hashBytes
+      , KnownNat len
+      , Na.CRYPTO_GENERICHASH_BYTES_MIN <= len
+      , len <= Na.CRYPTO_GENERICHASH_BYTES_MAX
+      )
+  => pt  -- ^ Message to hash
+  -> Maybe key -- ^ Hash key
```

Comment:
Is there any reason particular you chose this order of arguments (apart from it being such upstream)? I think, the key should go first so that you can have `simple_blake2b = blake2b Nothing`.

Reply:
Convenience of implementation and upstream convention, yes. I don't mind changing that.

Reply:
👌🏻
I just checked other functions in the library and, yes, looks like we are consistently reversing the order of arguments so make sure keys go first, so it’s better to do so here too.